File: backend/sdk/csharp/TradingClient.cs
/**
 * AI Trading Evolution - C# SDK
 * ==============================
 * Official C# client library for AI Trading Evolution API
 * 
 * Author: MiniMax Agent
 * Version: 1.0.0
 * Date: 2025-11-04
 * 
 * Usage:
 *   var client = new TradingClient("http://localhost:8000", "your-api-key");
 *   
 *   // Get market data
 *   var data = await client.Market.GetDataAsync("BTC/USDT");
 *   
 *   // Execute strategy
 *   var signal = await client.Strategy.ExecuteAsync("grid", "BTC/USDT");
 *   
 *   // Run analytics
 *   var sentiment = await client.Analytics.SentimentAsync("BTC/USDT");
 */

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace AITrading.SDK
{
    // =========================================================================
    // Exceptions
    // =========================================================================

    public class APIException : Exception
    {
        public int StatusCode { get; }

        public APIException(string message, int statusCode) : base(message)
        {
            StatusCode = statusCode;
        }
    }

    // =========================================================================
    // Models
    // =========================================================================

    public class StrategyResponse
    {
        public string StrategyName { get; set; }
        public string Symbol { get; set; }
        public string Signal { get; set; }
        public double Confidence { get; set; }
        public double Price { get; set; }
        public double? EntryPrice { get; set; }
        public double? StopLoss { get; set; }
        public double? TakeProfit { get; set; }
        public string Timestamp { get; set; }
        public Dictionary<string, object> Metadata { get; set; }
    }

    public class AnalyticsResponse
    {
        public string AnalysisType { get; set; }
        public Dictionary<string, object> Result { get; set; }
        public string Timestamp { get; set; }
    }

    // =========================================================================
    // Base Client
    // =========================================================================

    internal class BaseClient
    {
        private readonly string _baseUrl;
        private readonly string _apiKey;
        private readonly HttpClient _httpClient;
        private readonly JsonSerializerOptions _jsonOptions;

        public BaseClient(string baseUrl, string apiKey)
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _apiKey = apiKey;
            _httpClient = new HttpClient
            {
                Timeout = TimeSpan.FromSeconds(30)
            };

            if (!string.IsNullOrEmpty(apiKey))
            {
                _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {apiKey}");
            }

            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            };
        }

        public async Task<T> GetAsync<T>(string endpoint)
        {
            var url = $"{_baseUrl}{endpoint}";
            var response = await _httpClient.GetAsync(url);

            if (!response.IsSuccessStatusCode)
            {
                var error = await response.Content.ReadAsStringAsync();
                throw new APIException($"API Error {response.StatusCode}", (int)response.StatusCode);
            }

            var content = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<T>(content, _jsonOptions);
        }

        public async Task<T> PostAsync<T>(string endpoint, object data)
        {
            var url = $"{_baseUrl}{endpoint}";
            var json = JsonSerializer.Serialize(data, _jsonOptions);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _httpClient.PostAsync(url, content);

            if (!response.IsSuccessStatusCode)
            {
                var error = await response.Content.ReadAsStringAsync();
                throw new APIException($"API Error {response.StatusCode}", (int)response.StatusCode);
            }

            var responseContent = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<T>(responseContent, _jsonOptions);
        }
    }

    // =========================================================================
    // Market API
    // =========================================================================

    public class MarketAPI
    {
        private readonly BaseClient _client;

        internal MarketAPI(BaseClient client)
        {
            _client = client;
        }

        /// <summary>
        /// Get market data
        /// </summary>
        public async Task<object> GetDataAsync(
            string symbol,
            string marketType = "crypto",
            string timeframe = "1h",
            int limit = 100)
        {
            var request = new
            {
                symbol,
                market_type = marketType,
                timeframe,
                limit
            };

            return await _client.PostAsync<object>("/api/v1/market/data", request);
        }

        /// <summary>
        /// Get list of symbols
        /// </summary>
        public async Task<object> ListSymbolsAsync(string marketType = "crypto")
        {
            return await _client.GetAsync<object>($"/api/v1/market/symbols?market_type={marketType}");
        }
    }

    // =========================================================================
    // Strategy API
    // =========================================================================

    public class StrategyAPI
    {
        private readonly BaseClient _client;

        internal StrategyAPI(BaseClient client)
        {
            _client = client;
        }

        /// <summary>
        /// Execute trading strategy
        /// </summary>
        public async Task<StrategyResponse> ExecuteAsync(
            string strategyName,
            string symbol,
            string timeframe = "1h",
            Dictionary<string, object> parameters = null)
        {
            var request = new
            {
                strategy_name = strategyName,
                symbol,
                timeframe,
                parameters = parameters ?? new Dictionary<string, object>()
            };

            return await _client.PostAsync<StrategyResponse>("/api/v1/strategy/execute", request);
        }

        /// <summary>
        /// Get list of strategies
        /// </summary>
        public async Task<object> ListAsync()
        {
            return await _client.GetAsync<object>("/api/v1/strategy/list");
        }
    }

    // =========================================================================
    // Analytics API
    // =========================================================================

    public class AnalyticsAPI
    {
        private readonly BaseClient _client;

        internal AnalyticsAPI(BaseClient client)
        {
            _client = client;
        }

        /// <summary>
        /// Run analytics
        /// </summary>
        public async Task<AnalyticsResponse> AnalyzeAsync(
            string analysisType,
            string symbol = null,
            Dictionary<string, object> parameters = null)
        {
            var request = new
            {
                analysis_type = analysisType,
                symbol,
                parameters = parameters ?? new Dictionary<string, object>()
            };

            return await _client.PostAsync<AnalyticsResponse>("/api/v1/analytics/analyze", request);
        }

        /// <summary>
        /// Get sentiment analysis
        /// </summary>
        public async Task<AnalyticsResponse> SentimentAsync(string symbol)
        {
            return await AnalyzeAsync("sentiment", symbol);
        }

        /// <summary>
        /// Get risk scoring
        /// </summary>
        public async Task<AnalyticsResponse> RiskScoringAsync(string symbol)
        {
            return await AnalyzeAsync("risk_scoring", symbol);
        }

        /// <summary>
        /// Get list of analytics types
        /// </summary>
        public async Task<object> ListTypesAsync()
        {
            return await _client.GetAsync<object>("/api/v1/analytics/types");
        }
    }

    // =========================================================================
    // Main Trading Client
    // =========================================================================

    /// <summary>
    /// Main AI Trading Evolution client
    /// </summary>
    public class TradingClient
    {
        private readonly BaseClient _baseClient;

        public MarketAPI Market { get; }
        public StrategyAPI Strategy { get; }
        public AnalyticsAPI Analytics { get; }

        /// <summary>
        /// Create new TradingClient
        /// </summary>
        /// <param name="baseUrl">API base URL (e.g., "http://localhost:8000")</param>
        /// <param name="apiKey">API key (optional)</param>
        public TradingClient(string baseUrl, string apiKey = null)
        {
            _baseClient = new BaseClient(baseUrl, apiKey);

            Market = new MarketAPI(_baseClient);
            Strategy = new StrategyAPI(_baseClient);
            Analytics = new AnalyticsAPI(_baseClient);
        }

        /// <summary>
        /// Get API health status
        /// </summary>
        public async Task<object> HealthAsync()
        {
            return await _baseClient.GetAsync<object>("/health");
        }

        /// <summary>
        /// Get performance metrics
        /// </summary>
        public async Task<object> MetricsAsync()
        {
            return await _baseClient.GetAsync<object>("/metrics");
        }
    }
}

File: backend/code/sdk/go/trading_client.go
/**
 * AI Trading Evolution - Go SDK
 * ==============================
 * Official Go client library for AI Trading Evolution API
 * 
 * Author: MiniMax Agent
 * Version: 1.0.0
 * Date: 2025-11-04
 * 
 * Usage:
 *   client := trading.NewClient("http://localhost:8000", "your-api-key")
 *   
 *   // Get market data
 *   data, err := client.Market.GetData("BTC/USDT", nil)
 *   
 *   // Execute strategy
 *   signal, err := client.Strategy.Execute("grid", "BTC/USDT", nil)
 *   
 *   // Run analytics
 *   sentiment, err := client.Analytics.Sentiment("BTC/USDT")
 */

package trading

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// =============================================================================
// Types
// =============================================================================

// APIError represents an API error
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("API Error %d: %s", e.StatusCode, e.Message)
}

// MarketDataRequest represents market data request
type MarketDataRequest struct {
	Symbol     string                 `json:"symbol"`
	MarketType string                 `json:"market_type"`
	Timeframe  string                 `json:"timeframe"`
	Limit      int                    `json:"limit"`
}

// StrategyRequest represents strategy execution request
type StrategyRequest struct {
	StrategyName string                 `json:"strategy_name"`
	Symbol       string                 `json:"symbol"`
	Timeframe    string                 `json:"timeframe"`
	Parameters   map[string]interface{} `json:"parameters"`
}

// AnalyticsRequest represents analytics request
type AnalyticsRequest struct {
	AnalysisType string                 `json:"analysis_type"`
	Symbol       string                 `json:"symbol,omitempty"`
	Parameters   map[string]interface{} `json:"parameters"`
}

// StrategyResponse represents strategy execution response
type StrategyResponse struct {
	StrategyName string                 `json:"strategy_name"`
	Symbol       string                 `json:"symbol"`
	Signal       string                 `json:"signal"`
	Confidence   float64                `json:"confidence"`
	Price        float64                `json:"price"`
	EntryPrice   *float64               `json:"entry_price,omitempty"`
	StopLoss     *float64               `json:"stop_loss,omitempty"`
	TakeProfit   *float64               `json:"take_profit,omitempty"`
	Timestamp    string                 `json:"timestamp"`
	Metadata     map[string]interface{} `json:"metadata"`
}

// AnalyticsResponse represents analytics response
type AnalyticsResponse struct {
	AnalysisType string                 `json:"analysis_type"`
	Result       map[string]interface{} `json:"result"`
	Timestamp    string                 `json:"timestamp"`
}

// =============================================================================
// Base Client
// =============================================================================

// BaseClient handles HTTP requests
type BaseClient struct {
	baseURL    string
	apiKey     string
	httpClient *http.Client
}

// NewBaseClient creates a new base client
func NewBaseClient(baseURL, apiKey string) *BaseClient {
	return &BaseClient{
		baseURL: baseURL,
		apiKey:  apiKey,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// request makes HTTP request
func (c *BaseClient) request(method, endpoint string, body interface{}, result interface{}) error {
	url := c.baseURL + endpoint

	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		var errorResponse map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResponse)
		message := "Unknown error"
		if msg, ok := errorResponse["error"].(string); ok {
			message = msg
		}
		return &APIError{
			StatusCode: resp.StatusCode,
			Message:    message,
		}
	}

	if result != nil {
		return json.NewDecoder(resp.Body).Decode(result)
	}

	return nil
}

// Get makes GET request
func (c *BaseClient) Get(endpoint string, result interface{}) error {
	return c.request("GET", endpoint, nil, result)
}

// Post makes POST request
func (c *BaseClient) Post(endpoint string, body, result interface{}) error {
	return c.request("POST", endpoint, body, result)
}

// =============================================================================
// Market API
// =============================================================================

// MarketAPI handles market data operations
type MarketAPI struct {
	client *BaseClient
}

// GetData retrieves market data
func (m *MarketAPI) GetData(symbol string, options *MarketDataRequest) (map[string]interface{}, error) {
	if options == nil {
		options = &MarketDataRequest{
			Symbol:     symbol,
			MarketType: "crypto",
			Timeframe:  "1h",
			Limit:      100,
		}
	} else {
		options.Symbol = symbol
		if options.MarketType == "" {
			options.MarketType = "crypto"
		}
		if options.Timeframe == "" {
			options.Timeframe = "1h"
		}
		if options.Limit == 0 {
			options.Limit = 100
		}
	}

	var result map[string]interface{}
	err := m.client.Post("/api/v1/market/data", options, &result)
	return result, err
}

// ListSymbols retrieves available symbols
func (m *MarketAPI) ListSymbols(marketType string) (map[string]interface{}, error) {
	var result map[string]interface{}
	err := m.client.Get("/api/v1/market/symbols?market_type="+marketType, &result)
	return result, err
}

// =============================================================================
// Strategy API
// =============================================================================

// StrategyAPI handles strategy operations
type StrategyAPI struct {
	client *BaseClient
}

// Execute executes trading strategy
func (s *StrategyAPI) Execute(strategyName, symbol string, options *StrategyRequest) (*StrategyResponse, error) {
	if options == nil {
		options = &StrategyRequest{
			StrategyName: strategyName,
			Symbol:       symbol,
			Timeframe:    "1h",
			Parameters:   make(map[string]interface{}),
		}
	} else {
		options.StrategyName = strategyName
		options.Symbol = symbol
		if options.Timeframe == "" {
			options.Timeframe = "1h"
		}
		if options.Parameters == nil {
			options.Parameters = make(map[string]interface{})
		}
	}

	var result StrategyResponse
	err := s.client.Post("/api/v1/strategy/execute", options, &result)
	return &result, err
}

// List retrieves available strategies
func (s *StrategyAPI) List() (map[string]interface{}, error) {
	var result map[string]interface{}
	err := s.client.Get("/api/v1/strategy/list", &result)
	return result, err
}

// =============================================================================
// Analytics API
// =============================================================================

// AnalyticsAPI handles analytics operations
type AnalyticsAPI struct {
	client *BaseClient
}

// Analyze runs analytics
func (a *AnalyticsAPI) Analyze(analysisType string, symbol *string, parameters map[string]interface{}) (*AnalyticsResponse, error) {
	req := &AnalyticsRequest{
		AnalysisType: analysisType,
		Parameters:   parameters,
	}
	if symbol != nil {
		req.Symbol = *symbol
	}
	if req.Parameters == nil {
		req.Parameters = make(map[string]interface{})
	}

	var result AnalyticsResponse
	err := a.client.Post("/api/v1/analytics/analyze", req, &result)
	return &result, err
}

// Sentiment retrieves sentiment analysis
func (a *AnalyticsAPI) Sentiment(symbol string) (*AnalyticsResponse, error) {
	return a.Analyze("sentiment", &symbol, nil)
}

// RiskScoring retrieves risk scoring
func (a *AnalyticsAPI) RiskScoring(symbol string) (*AnalyticsResponse, error) {
	return a.Analyze("risk_scoring", &symbol, nil)
}

// ListTypes retrieves available analytics types
func (a *AnalyticsAPI) ListTypes() (map[string]interface{}, error) {
	var result map[string]interface{}
	err := a.client.Get("/api/v1/analytics/types", &result)
	return result, err
}

// =============================================================================
// Main Trading Client
// =============================================================================

// Client is the main AI Trading Evolution client
type Client struct {
	baseClient *BaseClient

	Market    *MarketAPI
	Strategy  *StrategyAPI
	Analytics *AnalyticsAPI
}

// NewClient creates a new trading client
func NewClient(baseURL, apiKey string) *Client {
	baseClient := NewBaseClient(baseURL, apiKey)

	return &Client{
		baseClient: baseClient,
		Market:     &MarketAPI{client: baseClient},
		Strategy:   &StrategyAPI{client: baseClient},
		Analytics:  &AnalyticsAPI{client: baseClient},
	}
}

// Health retrieves API health status
func (c *Client) Health() (map[string]interface{}, error) {
	var result map[string]interface{}
	err := c.baseClient.Get("/health", &result)
	return result, err
}

// Metrics retrieves performance metrics
func (c *Client) Metrics() (map[string]interface{}, error) {
	var result map[string]interface{}
	err := c.baseClient.Get("/metrics", &result)
	return result, err
}
